File: solutions/cpp/generate-parentheses.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <random>
#include <stack>
#include <string>
#include <vector>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::random_device;
using std::stack;
using std::stoi;
using std::string;
using std::uniform_int_distribution;
using std::vector;

void GenerateParenthesesHelper(int remained, int left_parens, string* s,
                               vector<string>* result);

// @include
vector<string> GenerateParentheses(int n) {
  string s;
  vector<string> result;
  GenerateParenthesesHelper(2 * n, 0, &s, &result);
  return result;
}

void GenerateParenthesesHelper(int remained, int left_parens, string* s,
                               vector<string>* result) {
  if (!remained) {
    result->emplace_back(*s);
    return;
  }

  if (left_parens < remained) {  // Able to insert '('.
    s->push_back('(');
    GenerateParenthesesHelper(remained - 1, left_parens + 1, s, result);
    s->pop_back();
  }
  if (left_parens > 0) {  // Able to insert ')'.
    s->push_back(')');
    GenerateParenthesesHelper(remained - 1, left_parens - 1, s, result);
    s->pop_back();
  }
}
// @exclude

void SmallTest() {
  auto result = GenerateParentheses(1);
  assert(result.size() == 1 && !result.front().compare("()"));
  result = GenerateParentheses(2);
  assert(result.size() == 2 &&
         ((!result.front().compare("(())") && !result.back().compare("()()")) ||
          (!result.front().compare("()()") && !result.back().compare("(())"))));
}

int main(int argc, char** argv) {
  SmallTest();
  default_random_engine gen((random_device())());
  int n;
  if (argc == 2) {
    n = stoi(argv[1]);
  } else {
    uniform_int_distribution<int> dis_n(0, 10);
    n = dis_n(gen);
  }
  cout << "n = " << n << endl;
  auto result = GenerateParentheses(n);
  for (const string& s : result) {
    cout << s << " ";
  }
  cout << endl;
  return 0;
}

File: solutions/python/generate_parentheses.py
import sys
import random


# @include
def generate_balanced_parentheses(num_pairs):
    def directed_generate_balanced_parentheses(num_left_parens_needed,
                                               num_right_parens_needed,
                                               valid_prefix,
                                               result=[]):
        if num_left_parens_needed > 0:  # Able to insert '('.
            directed_generate_balanced_parentheses(num_left_parens_needed - 1,
                                                   num_right_parens_needed,
                                                   valid_prefix + '(')
        if num_left_parens_needed < num_right_parens_needed:
            # Able to insert ')'.
            directed_generate_balanced_parentheses(num_left_parens_needed,
                                                   num_right_parens_needed - 1,
                                                   valid_prefix + ')')
        if not num_right_parens_needed:
            result.append(valid_prefix)
        return result

    return directed_generate_balanced_parentheses(num_pairs, num_pairs, '')
# @exclude


def generate_balanced_parentheses_pythonic(num_pairs, num_left_open=0):
    if not num_pairs:
        return [')' * num_left_open]
    if not num_left_open:
        return [
            '(' + p
            for p in generate_balanced_parentheses_pythonic(num_pairs - 1,
                                                            num_left_open + 1)
        ]
    else:
        return ([
            '(' + p
            for p in generate_balanced_parentheses_pythonic(num_pairs - 1,
                                                            num_left_open + 1)
        ] + [
            ')' + p
            for p in generate_balanced_parentheses_pythonic(num_pairs - 1,
                                                            num_left_open - 1)
        ])


def small_test():
    assert generate_balanced_parentheses(1) == ['()']
    result = generate_balanced_parentheses(2)
    assert result == ['(())', '()()'] or result == ['()()', '(())']


def main():
    small_test()
    num_pairs = int(sys.argv[1]) if len(sys.argv) == 2 else random.randint(0,
                                                                           10)
    print('num_pairs =', num_pairs)
    result = generate_balanced_parentheses(num_pairs)
    print(*result)


if __name__ == '__main__':
    main()
